nio/router/diagnostic.py

```diff
@@ -0,0 +1,51 @@
+from datetime import timedelta
+from threading import RLock
+
+from nio.modules.scheduler.job import Job
+from nio.util.runner import Runner
+
+
+class DiagnosticManager(Runner):
+
+    def __init__(self):
+        super().__init__()
+
+        self._interval = None
+        self._mgmt_signal_handler = None
+        self._job = None
+
+        self._signals_lock = RLock()
+        self._signals = []
+
+    def configure(self, context):
+        self._interval = \
+            context.settings.get("diagnostic_interval", 30)
+        self._mgmt_signal_handler = context.mgmt_signal_handler
+
+    def start(self):
+        super().start()
+
+        # create job to periodically send updates
+        if self._interval > 0:
+            self._job = \
+                Job(self._send_diagnostic,
+                    timedelta(seconds=self._interval),
+                    True)
+
+    def stop(self):
+        if self._job:
+            self._job.cancel()
+
+        # Send what we have saved
+        self._send_diagnostic()
+        super().stop()
+
+    def on_signal(self, signal):
+        with self._signals_lock:
+            self._signals.append(signal)
```

Comment:
I would actually rather have this diagnostic manager consolidate these signals together. In other words, if I've set up the diagnostic interval to notify every minute, I wouldn't want to still get a ton of signals sent about everything that happened in that minute. Given this is only diagnostic data we don't need that level of granularity. Instead, I'd rather have one signal sent for every output-input block combination with the number of signals sent between them during that minute.

Reply:
good catch, makes good sense